### DISCUSSION TITLE
How does a Perpetual License detect whether end-user is attempting to install a new version.

### DISCUSSION THREAD

**Discussion by ManlyBoy** [keygen-sh/community#105] [Q&A] [answered]
" but the twist is that the customer will also be allowed to download (and use) any product version released within their license's 1 year window, even after the license expires."

How is this achieved? What do I need to do to implement this validation within my application?

**Comment by ezekg** [comment]
To enable a perpetual fallback license, you can use [the `MAINTAIN_ACCESS` expiration strategy](https://keygen.sh/docs/api/policies/#policies-object-attrs-expirationStrategy).

It uses the release's creation date to determine if the release is accessible to the license. More simply, if `release.created_at < license.expiry`, then the download is allowed. You do not need to implement additional validation on your end, but you can use [the `scope.version` validation parameter](https://keygen.sh/docs/api/licenses/#licenses-validate-meta-scope.version) to track each license's installed version.

**Reply by ManlyBoy** [reply]
"To enable a perpetual fallback license, you can use [the MAINTAIN_ACCESS expiration strategy](https://keygen.sh/docs/api/policies/#policies-object-attrs-expirationStrategy)."
Yes, I read this with great interest as I think it's great and exactly what I am looking for. 

Probably this may seem stupid but I am very new to keygen but exactly what will be downloaded if it is allowed.
I presume there is nothing stopping someone from downloading the upgrade or new version but I assume that when they go to install their current licence the something won't be allowed to be downloaded let alone installed. If so, I need to check this in my installer before they try to install the program else by the time they realize the licence is no good they may have already overwritten their current installation. ditto with my automatic updater.

**Reply by ezekg** [reply, marked as answer]
You'll need to distribute your application using Keygen's API, via [the release](https://keygen.sh/docs/api/releases/) and [artifact](https://keygen.sh/docs/api/artifacts/) resources. So in this case, your application will be what's being downloaded. Using our API for distribution effectively gate your downloads and upgrades to licensed users. They'd use their license key to authenticate with our API and download/upgrade files. This does require the license to be created before the download, but there are ways around that via [the product's distribution strategy](https://keygen.sh/docs/api/products/#products-object-attrs-distributionStrategy).

But yes, somebody could still obtain an application binary out-of-band (e.g. from a friend), so you should also assert that the given license can utilize a particular release by using the `meta.scope.version` validation scope mentioned in my previous comment.

When the license isn't allowed to access a particular release, a `VERSION_SCOPE_MISMATCH` validation code will be returned, and the license will be considered invalid. You can use this as an indicator at runtime to allow/disallow application usage.

**Comment by ManlyBoy** [comment]
Thank you for this excellent explanation. I guessed I was missing out on some information that everyone else seemed to know and take for granted.
